lambdas/app.js

```diff
@@ -38,8 +37,23 @@ const originRequest = async (event, context) => {
     });
 };
 
-const originResponse = async (event) => {
-  const { response } = event.Records[0].cf;
+const originResponse = async (event, context) => {
+  const { request, response } = event.Records[0].cf; 
+  if (['404', '403'].includes(response.status)) {
+    const host = Utils.getHost(request);
+    const sitePath = getSitePath(request);
+    const params = getBuildQueryParams(host, sitePath, context.functionName);
+    const build = await getSiteItem(params);
+    const { Settings: { spa } } = build;
+    if (spa) {
+      const errorDocPath = [sitePath, 'index.html'].join('/');
```

Comment:
This may not be exactly what we want. I *think* what we want to do is only return index.html for 404s for html requests. That is, requests for assets that don't exist, such as images, css, js, etc..., should still return a 404.

Given this, one thing to think about is if we can actually do this up front. For example, instead of waiting for a 404 on an html request, if we know the site has spa routing configured, we can just rewrite the request to always return index.html. This would eliminate the extra request to S3. I'm still thinking on whether this would be a good idea or not....

Logically, this is slightly different in that if we wait for a 404, a site could have other html files that are not part of the SPA that could be returned, but I don't think anyone would actually do this 🤷 .

Reply:
@davemcorwin  - what you say makes sense to me and could very well be right.  however, the solution as-is is consistent with all the blog posts I have seen on handling SPA routing in Lambda@Edgethis where there is not a distinction made on the file type requested. this may have to be tried first to determine if all the solutions I've read have it wrong.

if we could do this up front that'd be ideal and perhaps something we can iterate on. this is consistent with the blog posts i have found on SPA in Lambda@Edge and essentially mimics the behavior off handling SPA routing in Cloudfront by modifying the errorDocument to be index.html.

Reply:
The description of the `opt.ignores` describes their logic in this node static file server: https://github.com/lukeed/sirv/tree/master/packages/sirv#optssingle

Reply:
> By default, any asset-like path (URLs that end with an extension) will be ignored. This means that, for example, if /foobar.jpg is not found, a 404 response is sent instead of the index.html fallback.

> Additionally, any /.well-known/* pathname ignores the fallback – as do all other dotfile requests when opts.dotfiles is enabled.